# Atividade Prática — Aula 4: Análise de Dados e Construção de Indicadores (KPIs)

Esta atividade foi construída com base na Aula 4, que marca a transição **do dado limpo para a tomada de decisão estratégica**, com foco em **KPIs, métricas derivadas, dimensões de análise, rankings e insights executivos**.

## Ideia central da aula
Antes do código, vem a **pergunta de negócio**.  
Depois, definimos:
1. a **métrica**
2. a **dimensão**
3. a **agregação**
4. e só então construímos a resposta em tabela e em texto.

## Regras da atividade
- O notebook **orienta**, mas você deve **escrever os códigos**.
- Sempre que gerar uma tabela agregada, escreva **1 ou 2 frases interpretando o resultado**.
- O objetivo não é apenas calcular números, mas comunicar decisões.

## Dataset da atividade
Arquivo: `vendas_brasil_aula4_kpis.csv`


## 1. Preparação do ambiente

Importe as bibliotecas necessárias para a atividade.

**Sugestão:**  
- `pandas`
- `numpy`

Se desejar, você também pode usar uma biblioteca de visualização mais adiante, mas o foco desta aula é **tabela agregada + insight**.


In [10]:
# Escreva aqui suas importações
import pandas as pd
import numpy as np
from pathlib import Path


## 2. Leitura da base

Leia o arquivo `vendas_brasil_aula4_kpis.csv` em um DataFrame chamado `df`.

Depois:
1. exiba as primeiras linhas
2. veja o tamanho da base
3. inspecione os tipos das colunas
4. confirme se `data_venda` está pronta para análises temporais


In [11]:
# Leia o CSV e faça a inspeção inicial
csv_path = Path('vendas_brasil_aula4_kpis.csv')
df = pd.read_csv(csv_path)

df['data_venda'] = pd.to_datetime(df['data_venda'], errors='coerce')

display(df.head())
print('Formato da base (linhas, colunas):', df.shape)
print('\nTipos das colunas:')
print(df.dtypes)
print('\nValores nulos em data_venda:', df['data_venda'].isna().sum())


,pedido_id,data_venda,uf,canal_venda,categoria,produto,quantidade_pedidos,preco_unitario,receita,lucro
0,5000,2025-02-14,RJ,Marketplace,Periféricos,Mouse Gamer,2,237.65,475.30,239.08
1,5001,2025-02-17,PR,Online,Informática,Monitor 27,4,1431.79,5727.16,1769.48
2,5002,2025-07-22,RJ,Marketplace,Informática,Monitor 27,5,1446.22,7231.10,2449.85
3,5003,2025-12-08,SP,Marketplace,Periféricos,Teclado Mecânico,7,399.84,2798.88,1356.67
4,5004,2025-01-04,GO,Marketplace,Telefonia,Smartphone X,4,3134.85,12539.40,3656.84


Formato da base (linhas, colunas): (420, 10)

Tipos das colunas:
pedido_id                      int64
data_venda            datetime64[us]
uf                               str
canal_venda                      str
categoria                        str
produto                          str
quantidade_pedidos             int64
preco_unitario               float64
receita                      float64
lucro                        float64
dtype: object

Valores nulos em data_venda: 0


## 3. Perguntas antes do código

A aula enfatiza que a maior armadilha é começar com `groupby()` sem saber o que se quer responder.

### Tarefa
Antes de programar, escreva em markdown respostas curtas para estas perguntas:

1. Qual canal de venda é o mais eficiente?
2. Qual categoria parece mais rentável?
3. Quais produtos são os campeões por lucro?
4. O negócio está crescendo ao longo do tempo?
5. Quais KPIs globais resumem melhor a operação?

### Respostas iniciais
- **Canal mais eficiente:** Online, pois apresenta a maior margem de lucro entre os canais e um ticket médio mais alto. Marketplace lidera em receita, mas não em eficiência.
- **Categoria mais rentável:** Telefonia, por concentrar a maior receita e o maior lucro total; Informática também tem peso relevante no resultado.
- **Campeões por lucro:** Notebook Pro, Smartphone X e Tablet Plus se destacam como os produtos mais lucrativos.
- **Evolução no tempo:** Não há crescimento linear mês a mês, mas a base mostra aceleração no segundo semestre, com pico forte no fim do ano, especialmente em novembro e dezembro.
- **KPIs globais mais úteis:** Receita total, lucro total, margem de lucro e ticket médio. Juntos, eles mostram tamanho, eficiência e valor médio das vendas.

Depois disso, avance para o código.


## 4. KPIs globais da operação

Com base nas “4 Métricas de Ouro do Varejo” apresentadas na aula, calcule os seguintes KPIs globais da base inteira:

- **Receita Total**
- **Lucro Total**
- **Margem de Lucro (%)**
- **Ticket Médio**

### Orientação
- Pense primeiro na fórmula de cada KPI
- Escreva os cálculos em células separadas ou em uma pequena tabela-resumo
- Se quiser, formate os resultados para facilitar a leitura


In [1]:
import pandas as pd
from pathlib import Path

csv_path = Path('vendas_brasil_aula4_kpis.csv')

if 'df' not in globals():
    df = pd.read_csv(csv_path)

if 'data_venda' in df.columns:
    df['data_venda'] = pd.to_datetime(df['data_venda'], errors='coerce')

# Calcule aqui os 4 KPIs globais
receita_total = df['receita'].sum()
lucro_total = df['lucro'].sum()
margem_lucro = lucro_total / receita_total if receita_total else 0
ticket_medio = receita_total / df['pedido_id'].nunique() if 'pedido_id' in df.columns else df['receita'].mean()

kpis_globais = pd.DataFrame({
    'KPI': ['Receita Total', 'Lucro Total', 'Margem de Lucro', 'Ticket Médio'],
    'Valor': [receita_total, lucro_total, margem_lucro, ticket_medio]
})

kpis_globais['Valor'] = [
    f'R$ {receita_total:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'),
    f'R$ {lucro_total:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'),
    f'{margem_lucro:.2%}'.replace('.', ','),
    f'R$ {ticket_medio:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.')
]

kpis_globais


,KPI,Valor
0,Receita Total,"R$ 1.921.492,22"
1,Lucro Total,"R$ 544.496,15"
2,Margem de Lucro,"28,34%"
3,Ticket Médio,"R$ 4.574,98"


### Interpretação
O negócio parece de porte relevante, com receita total acima de R$ 1,9 milhão e lucro total superior a R$ 544 mil.
A margem de lucro em torno de 28% indica uma operação saudável, com boa conversão da receita em resultado.
O ticket médio próximo de R$ 4,6 mil sugere vendas de valor alto por pedido, o que reforça um perfil de operação mais consultiva e menos pulverizada.
Em resumo, a receita está sendo bem convertida em lucro, mas vale acompanhar a eficiência por canal e por categoria para sustentar esse desempenho.


## 5. Métricas derivadas

A aula mostra que **somar valores não basta**.  
Precisamos criar métricas derivadas para avaliar eficiência e comportamento.

### Tarefa
Crie, quando fizer sentido:
- uma coluna de **margem_lucro** (`lucro / receita`)
- uma lógica para **ticket_medio** da operação ou por grupo

### Reflexão
Explique:
1. Por que receita alta não significa automaticamente melhor desempenho?
2. Em que situação a margem ajuda mais do que a receita bruta?


In [2]:
# Crie aqui as métricas derivadas
df['margem_lucro'] = df['lucro'] / df['receita']
ticket_medio_operacao = df['receita'].sum() / df['pedido_id'].nunique() if 'pedido_id' in df.columns else df['receita'].mean()

ticket_medio_por_canal = (
    df.groupby('canal_venda')['receita'].sum() / df.groupby('canal_venda')['pedido_id'].nunique()
) if 'pedido_id' in df.columns else df.groupby('canal_venda')['receita'].mean()

metricas_derivadas = pd.DataFrame({
    'margem_lucro_media': [df['margem_lucro'].mean()],
    'ticket_medio_operacao': [ticket_medio_operacao]
})

metricas_derivadas

ticket_medio_por_canal


canal_venda
Loja Física    4481.098322
Marketplace    4665.263050
Online         4580.096103
dtype: float64

### Reflexão respondida
1. Receita alta não significa automaticamente melhor desempenho porque uma operação pode vender muito e ainda assim ter custos elevados, reduzindo a eficiência do resultado final.
2. A margem ajuda mais do que a receita bruta quando é preciso comparar eficiência entre canais, categorias ou produtos, porque mostra quanto realmente sobra de cada venda.


## 6. Fatiando a realidade: análise por dimensão

A aula mostra que um KPI isolado não decide nada.  
É preciso perguntar: **como isso varia por canal, categoria, região e tempo?**

### Tarefa
Liste, em markdown, quais colunas do dataset representam:
- métrica
- dimensão
- tempo

### Resposta
- **Métrica:** `receita`, `lucro`, `quantidade_pedidos`, `preco_unitario`, `margem_lucro`.
- **Dimensão:** `canal_venda`, `categoria`, `produto`, `uf`.
- **Tempo:** `data_venda` (com derivados como `mes` e `trimestre`).


## 7. Análise por canal

### Pergunta de negócio
**Qual canal de venda é o mais eficiente?**

### Tarefa
Crie uma tabela agregada por `canal_venda` contendo, no mínimo:
- Receita Total
- Lucro Total
- Margem Média ou Margem Calculada
- Quantidade de pedidos
- Ticket Médio

### Depois responda
1. Qual canal gera mais receita?
2. Qual canal gera maior eficiência?
3. Se você fosse gestor, priorizaria volume ou margem?


In [3]:
import pandas as pd
from pathlib import Path

if 'df' not in globals():
    csv_path = Path('vendas_brasil_aula4_kpis.csv')
    df = pd.read_csv(csv_path)

tabela_canal = (
    df.groupby('canal_venda')
      .agg(Receita_Total=('receita', 'sum'),
           Lucro_Total=('lucro', 'sum'),
           Pedidos=('canal_venda', 'size'))
      .assign(Margem=lambda x: x['Lucro_Total'] / x['Receita_Total'],
              Ticket_Medio=lambda x: x['Receita_Total'] / x['Pedidos'])
      .reset_index()
      .sort_values('Receita_Total', ascending=False)
)

tabela_canal['Receita_Total'] = tabela_canal['Receita_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_canal['Lucro_Total'] = tabela_canal['Lucro_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_canal['Margem'] = tabela_canal['Margem'].map(lambda v: f'{v:.2%}'.replace('.', ','))
tabela_canal['Ticket_Medio'] = tabela_canal['Ticket_Medio'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

tabela_canal


,canal_venda,Receita_Total,Lucro_Total,Pedidos,Margem,Ticket_Medio
1,Marketplace,"R$ 657.802,09","R$ 185.802,57",141,"28,25%","R$ 4.665,26"
0,Loja Física,"R$ 640.797,06","R$ 174.583,16",143,"27,24%","R$ 4.481,10"
2,Online,"R$ 622.893,07","R$ 184.110,42",136,"29,56%","R$ 4.580,10"


### Insight obrigatório
O Marketplace concentra a maior receita, mas o canal Online apresenta a melhor margem e o maior ticket médio, indicando maior eficiência por venda.
Na prática, a decisão não deve priorizar apenas volume: faz sentido proteger o canal de maior faturamento e, ao mesmo tempo, ampliar ações no canal Online, que entrega melhor rentabilidade por pedido.


## 8. Análise por categoria

### Pergunta de negócio
**Qual categoria é mais lucrativa e qual parece mais eficiente?**

### Tarefa
Crie uma tabela agregada por `categoria` com:
- Receita Total
- Lucro Total
- Margem
- Ticket Médio

### Depois responda
1. A categoria de maior receita também é a de maior margem?
2. Existe alguma categoria que parece vender bem, mas ser pouco eficiente?


In [4]:
import pandas as pd
from pathlib import Path

if 'df' not in globals():
    csv_path = Path('vendas_brasil_aula4_kpis.csv')
    df = pd.read_csv(csv_path)

tabela_categoria = (
    df.groupby('categoria')
      .agg(Receita_Total=('receita', 'sum'),
           Lucro_Total=('lucro', 'sum'),
           Pedidos=('categoria', 'size'))
      .assign(Margem=lambda x: x['Lucro_Total'] / x['Receita_Total'],
              Ticket_Medio=lambda x: x['Receita_Total'] / x['Pedidos'])
      .reset_index()
      .sort_values('Lucro_Total', ascending=False)
)

tabela_categoria['Receita_Total'] = tabela_categoria['Receita_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_categoria['Lucro_Total'] = tabela_categoria['Lucro_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_categoria['Margem'] = tabela_categoria['Margem'].map(lambda v: f'{v:.2%}'.replace('.', ','))
tabela_categoria['Ticket_Medio'] = tabela_categoria['Ticket_Medio'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

tabela_categoria


,categoria,Receita_Total,Lucro_Total,Pedidos,Margem,Ticket_Medio
2,Telefonia,"R$ 866.292,45","R$ 219.975,58",109,"25,39%","R$ 7.947,64"
0,Informática,"R$ 805.376,44","R$ 219.090,15",93,"27,20%","R$ 8.659,96"
3,Áudio,"R$ 132.900,25","R$ 55.006,02",111,"41,39%","R$ 1.197,30"
1,Periféricos,"R$ 116.923,08","R$ 50.424,40",107,"43,13%","R$ 1.092,74"


### Resposta executiva
A categoria de maior receita e lucro é Telefonia, então ela é a principal responsável pelo resultado absoluto da operação.
Por outro lado, Periféricos e Áudio apresentam as maiores margens, o que indica mais eficiência; isso mostra que vender mais não é o mesmo que ser mais rentável.


## 9. Análise por região (UF)

### Pergunta de negócio
**Quais UFs concentram a operação?**

### Tarefa
Monte uma tabela por `uf` com:
- Receita Total
- Lucro Total
- Participação percentual da receita
- Ticket Médio

### Depois responda
1. Quais UFs concentram maior receita?
2. Existe concentração regional?
3. Se fosse necessário expandir ou reforçar logística, por onde você começaria?

### Resposta executiva
As UFs com maior receita são SP, RJ e PE, indicando que a operação está mais forte em estados de maior densidade econômica e consumo.
Existe concentração regional moderada, então a estratégia ideal é reforçar logística e disponibilidade primeiro nos estados líderes e, em paralelo, expandir de forma seletiva para UFs com potencial e participação menor.


In [5]:
tabela_uf = (
    df.groupby('uf')
      .agg(Receita_Total=('receita', 'sum'),
           Lucro_Total=('lucro', 'sum'),
           Pedidos=('uf', 'size'))
      .assign(Participacao_Receita=lambda x: x['Receita_Total'] / x['Receita_Total'].sum(),
              Ticket_Medio=lambda x: x['Receita_Total'] / x['Pedidos'])
      .reset_index()
      .sort_values('Receita_Total', ascending=False)
)

tabela_uf['Receita_Total'] = tabela_uf['Receita_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_uf['Lucro_Total'] = tabela_uf['Lucro_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
tabela_uf['Participacao_Receita'] = tabela_uf['Participacao_Receita'].map(lambda v: f'{v:.2%}'.replace('.', ','))
tabela_uf['Ticket_Medio'] = tabela_uf['Ticket_Medio'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

tabela_uf


,uf,Receita_Total,Lucro_Total,Pedidos,Participacao_Receita,Ticket_Medio
9,SP,"R$ 246.835,50","R$ 71.022,16",48,"12,85%","R$ 5.142,41"
6,RJ,"R$ 218.631,91","R$ 63.960,39",38,"11,38%","R$ 5.753,47"
4,PE,"R$ 208.457,44","R$ 58.183,95",55,"10,85%","R$ 3.790,14"
3,MG,"R$ 207.614,69","R$ 61.122,25",44,"10,80%","R$ 4.718,52"
8,SC,"R$ 198.143,47","R$ 60.272,32",40,"10,31%","R$ 4.953,59"
7,RS,"R$ 187.048,31","R$ 49.795,96",43,"9,73%","R$ 4.349,96"
1,ES,"R$ 181.113,70","R$ 49.838,07",39,"9,43%","R$ 4.643,94"
5,PR,"R$ 174.136,19","R$ 46.155,64",35,"9,06%","R$ 4.975,32"
0,BA,"R$ 157.082,99","R$ 42.117,30",42,"8,18%","R$ 3.740,07"
2,GO,"R$ 142.428,02","R$ 42.028,11",36,"7,41%","R$ 3.956,33"


## 10. O motor da análise: agregações (groupby)

Nos slides, a lógica central é:
- `receita -> sum`
- `margem -> mean` ou cálculo derivado
- `produto -> count`

### Tarefa
Escolha **duas dimensões diferentes** e construa tabelas agregadas mostrando que você entendeu essa lógica.

Exemplos:
- por canal
- por categoria
- por UF
- por mês


In [16]:
if 'margem_lucro' not in df.columns:
    df['margem_lucro'] = df['lucro'] / df['receita']

agregacao_canal = (
    df.groupby('canal_venda')
      .agg(Receita_Total=('receita', 'sum'),
           Margem_Media=('margem_lucro', 'mean'),
           Qtd_Produtos=('produto', 'count'))
      .sort_values('Receita_Total', ascending=False)
)

df['mes'] = pd.to_datetime(df['data_venda'], errors='coerce').dt.to_period('M').astype(str)
agregacao_mes = (
    df.groupby('mes')
      .agg(Receita_Total=('receita', 'sum'),
           Margem_Media=('margem_lucro', 'mean'),
           Qtd_Produtos=('produto', 'count'))
      .sort_index()
)

agregacao_canal_view = agregacao_canal.copy()
agregacao_canal_view['Receita_Total'] = agregacao_canal_view['Receita_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
agregacao_canal_view['Margem_Media'] = agregacao_canal_view['Margem_Media'].map(lambda v: f'{v:.2%}'.replace('.', ','))

agregacao_mes_view = agregacao_mes.copy()
agregacao_mes_view['Receita_Total'] = agregacao_mes_view['Receita_Total'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
agregacao_mes_view['Margem_Media'] = agregacao_mes_view['Margem_Media'].map(lambda v: f'{v:.2%}'.replace('.', ','))

agregacao_canal_view, agregacao_mes_view


(             Receita_Total Margem_Media  Qtd_Produtos
 canal_venda                                          
 Marketplace  R$ 657.802,09       34,72%           141
 Loja Física  R$ 640.797,06       34,49%           143
 Online       R$ 622.893,07       35,26%           136,
          Receita_Total Margem_Media  Qtd_Produtos
 mes                                              
 2025-01  R$ 161.571,89       35,04%            38
 2025-02  R$ 108.587,03       35,47%            31
 2025-03   R$ 87.902,22       34,28%            25
 2025-04  R$ 148.930,46       34,35%            43
 2025-05  R$ 106.355,17       33,35%            23
 2025-06  R$ 134.685,83       35,10%            33
 2025-07  R$ 196.651,96       34,28%            33
 2025-08  R$ 160.240,62       34,27%            37
 2025-09  R$ 109.291,03       35,98%            33
 2025-10  R$ 158.226,19       33,27%            34
 2025-11  R$ 325.537,31       33,44%            44
 2025-12  R$ 223.512,51       37,90%            46)

## 11. Campeões e vilões (Top N)

A aula destaca o poder dos rankings curtos e focados.

### Tarefa
Crie:
- Top 10 produtos por **lucro**
- Top 10 produtos por **receita**
- 5 categorias ou UFs com **pior margem** ou pior desempenho

### Perguntas
1. Quem são os “campeões”?
2. Quem são os “vilões”?
3. Que ação tática poderia ser tomada com base nisso?

### Respostas
Os campeões são Notebook Pro, Smartphone X e Tablet Plus, que lideram tanto em lucro quanto em receita.
Como vilões de eficiência, as categorias de maior receita (Telefonia e Informática) têm margens menores que Áudio e Periféricos.
Ação tática: preservar volume nos campeões e revisar preço, mix e custo nas categorias de menor margem para elevar rentabilidade sem perder faturamento.


In [13]:
top10_lucro = (
    df.groupby('produto', as_index=False)['lucro']
      .sum()
      .sort_values('lucro', ascending=False)
      .head(10)
)

top10_receita = (
    df.groupby('produto', as_index=False)['receita']
      .sum()
      .sort_values('receita', ascending=False)
      .head(10)
)

piores_categorias_margem = (
    df.groupby('categoria', as_index=False)
      .agg(receita=('receita', 'sum'), lucro=('lucro', 'sum'))
      .assign(margem=lambda x: x['lucro'] / x['receita'])
      .sort_values('margem', ascending=True)
      .head(5)
)

top10_lucro_view = top10_lucro.copy()
top10_lucro_view['lucro'] = top10_lucro_view['lucro'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

top10_receita_view = top10_receita.copy()
top10_receita_view['receita'] = top10_receita_view['receita'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

piores_categorias_margem_view = piores_categorias_margem.copy()
piores_categorias_margem_view['receita'] = piores_categorias_margem_view['receita'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
piores_categorias_margem_view['lucro'] = piores_categorias_margem_view['lucro'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
piores_categorias_margem_view['margem'] = piores_categorias_margem_view['margem'].map(lambda v: f'{v:.2%}'.replace('.', ','))

top10_lucro_view, top10_receita_view, piores_categorias_margem_view


(            produto          lucro
 4      Notebook Pro  R$ 153.094,84
 5      Smartphone X  R$ 118.198,25
 6       Tablet Plus  R$ 101.777,33
 2        Monitor 27   R$ 65.995,31
 7  Teclado Mecânico   R$ 33.939,07
 0      Caixa de Som   R$ 28.316,37
 1       Headset Pro   R$ 26.689,65
 3       Mouse Gamer   R$ 16.485,33,
             produto        receita
 4      Notebook Pro  R$ 594.132,74
 5      Smartphone X  R$ 489.572,08
 6       Tablet Plus  R$ 376.720,37
 2        Monitor 27  R$ 211.243,70
 7  Teclado Mecânico   R$ 79.973,75
 0      Caixa de Som   R$ 68.956,76
 1       Headset Pro   R$ 63.943,49
 3       Mouse Gamer   R$ 36.949,33,
      categoria        receita          lucro  margem
 2    Telefonia  R$ 866.292,45  R$ 219.975,58  25,39%
 0  Informática  R$ 805.376,44  R$ 219.090,15  27,20%
 3        Áudio  R$ 132.900,25   R$ 55.006,02  41,39%
 1  Periféricos  R$ 116.923,08   R$ 50.424,40  43,13%)

## 12. Análise temporal

A aula reforça que sem tempo a análise fica estática.

### Tarefa
1. Converta `data_venda` para data, se necessário
2. Crie colunas auxiliares como:
   - mês
   - trimestre (opcional)
3. Gere uma tabela agregada por mês com:
   - receita
   - lucro
   - ticket médio ou margem

### Depois responda
1. O lucro parece crescer ao longo do tempo?
2. O último trimestre foi melhor que o anterior?
3. Quais meses concentram as vendas de fim de ano?

### Respostas
O lucro não cresce de forma linear, mas há aceleração no segundo semestre com picos em novembro e dezembro.
O 4º trimestre supera o 3º trimestre em receita e lucro, sinalizando sazonalidade positiva de fim de ano.
Os meses que mais concentram vendas de fim de ano são novembro e dezembro.


In [17]:
if 'margem_lucro' not in df.columns:
    df['margem_lucro'] = df['lucro'] / df['receita']

df['data_venda'] = pd.to_datetime(df['data_venda'], errors='coerce')
df['mes'] = df['data_venda'].dt.to_period('M').astype(str)
df['trimestre'] = df['data_venda'].dt.to_period('Q').astype(str)

analise_mensal = (
    df.groupby('mes')
      .agg(receita=('receita', 'sum'),
           lucro=('lucro', 'sum'),
           margem=('margem_lucro', 'mean'))
      .reset_index()
      .sort_values('mes')
)

analise_trimestral = (
    df.groupby('trimestre')
      .agg(receita=('receita', 'sum'), lucro=('lucro', 'sum'))
      .reset_index()
      .sort_values('trimestre')
)

analise_mensal_view = analise_mensal.copy()
analise_mensal_view['receita'] = analise_mensal_view['receita'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
analise_mensal_view['lucro'] = analise_mensal_view['lucro'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
analise_mensal_view['margem'] = analise_mensal_view['margem'].map(lambda v: f'{v:.2%}'.replace('.', ','))

analise_trimestral_view = analise_trimestral.copy()
analise_trimestral_view['receita'] = analise_trimestral_view['receita'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))
analise_trimestral_view['lucro'] = analise_trimestral_view['lucro'].map(lambda v: f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.'))

analise_mensal_view, analise_trimestral_view


(        mes        receita         lucro  margem
 0   2025-01  R$ 161.571,89  R$ 46.515,14  35,04%
 1   2025-02  R$ 108.587,03  R$ 29.998,87  35,47%
 2   2025-03   R$ 87.902,22  R$ 23.923,66  34,28%
 3   2025-04  R$ 148.930,46  R$ 43.094,00  34,35%
 4   2025-05  R$ 106.355,17  R$ 30.334,32  33,35%
 5   2025-06  R$ 134.685,83  R$ 38.592,70  35,10%
 6   2025-07  R$ 196.651,96  R$ 57.913,44  34,28%
 7   2025-08  R$ 160.240,62  R$ 45.881,36  34,27%
 8   2025-09  R$ 109.291,03  R$ 31.938,44  35,98%
 9   2025-10  R$ 158.226,19  R$ 41.463,02  33,27%
 10  2025-11  R$ 325.537,31  R$ 87.763,18  33,44%
 11  2025-12  R$ 223.512,51  R$ 67.078,02  37,90%,
   trimestre        receita          lucro
 0    2025Q1  R$ 358.061,14  R$ 100.437,67
 1    2025Q2  R$ 389.971,46  R$ 112.021,02
 2    2025Q3  R$ 466.183,61  R$ 135.733,24
 3    2025Q4  R$ 707.276,01  R$ 196.304,22)

## 13. Fluxo de investigação do analista

A aula apresenta a sequência:
**Pergunta -> Colunas necessárias -> Tipo de resposta/gráfico**.

### Tarefa
Preencha, em markdown, pelo menos 3 exemplos no formato:

- **Pergunta de negócio:** ...
- **Colunas necessárias:** ...
- **Métrica/KPI:** ...
- **Tipo de resposta esperada:** tabela, ranking, tendência etc.

- **Pergunta de negócio:** Qual canal entrega maior eficiência com bom volume?
- **Colunas necessárias:** `canal_venda`, `receita`, `lucro`, `pedido_id`.
- **Métrica/KPI:** margem por canal, receita total e ticket médio.
- **Tipo de resposta esperada:** tabela comparativa por canal.

- **Pergunta de negócio:** Quais categorias sustentam o lucro da operação?
- **Colunas necessárias:** `categoria`, `receita`, `lucro`.
- **Métrica/KPI:** lucro total e margem por categoria.
- **Tipo de resposta esperada:** ranking por lucro e margem.

- **Pergunta de negócio:** O negócio melhora ou piora ao longo do tempo?
- **Colunas necessárias:** `data_venda`, `receita`, `lucro`, `margem_lucro`.
- **Métrica/KPI:** receita e lucro mensais, tendência trimestral.
- **Tipo de resposta esperada:** tabela temporal com tendência.


## 14. O insight exige palavras

Segundo a aula, o mercado valoriza quem comunica decisões, não apenas quem roda código.

### Tarefa
Escolha **duas tabelas agregadas** que você criou e escreva, para cada uma, **1 ou 2 frases de interpretação executiva**.

Exemplo de estilo:
- “O canal X concentra maior volume de receita, mas o canal Y apresenta margem superior.”
- “Recomenda-se revisar custos/expansão/priorização com base nesse resultado.”

### Interpretações
- **Tabela por canal:** Marketplace lidera em receita, mas Online tem melhor margem e ticket médio, então a estratégia deve equilibrar escala com eficiência.
- **Tabela por categoria:** Telefonia e Informática carregam o resultado absoluto, porém com margens inferiores às de Áudio e Periféricos; isso pede ajuste de mix e política comercial para melhorar rentabilidade.


## 15. Case Varejo Brasil — Entrega final

A aula propõe quatro entregas centrais para o gestor.

### Sua missão final
Organize o notebook para entregar:

1. **KPIs Globais**
2. **Tabelas por Dimensões Críticas**
   - por canal
   - por categoria
3. **Top 10 Produtos por Lucro**
4. **Conclusão final com pelo menos uma decisão tática**

### Importante
A conclusão final deve ser escrita em linguagem clara, direta e orientada à decisão.

### Conclusão final
A operação apresenta escala relevante e margem global saudável, com maior peso de receita em Marketplace e maior eficiência no canal Online.
As categorias Telefonia e Informática sustentam o volume, enquanto Áudio e Periféricos mostram maior margem relativa.
**Decisão tática:** manter investimento comercial nos campeões de receita/lucro e executar um plano de ganho de margem (preço, mix e custos) nas categorias de maior faturamento.


## 16. Desafio extra (opcional)

Monte uma pequena tabela executiva final com:
- Receita Total
- Lucro Total
- Margem
- Ticket Médio

e depois crie uma segunda tabela por canal com os mesmos indicadores.

Compare os resultados e explique:
- o KPI global esconde alguma diferença importante entre os canais?

### Comparação
O KPI global resume bem o desempenho médio, mas esconde diferenças de eficiência entre canais: o Online tem margem superior, enquanto o Marketplace lidera em volume.
Sem a visão por canal, a decisão poderia privilegiar apenas faturamento e perder oportunidades de aumentar rentabilidade.


In [18]:
resumo_executivo = pd.DataFrame({
    'KPI': ['Receita Total', 'Lucro Total', 'Margem', 'Ticket Médio'],
    'Valor': [
        df['receita'].sum(),
        df['lucro'].sum(),
        df['lucro'].sum() / df['receita'].sum(),
        df['receita'].sum() / df['pedido_id'].nunique()
    ]
})

resumo_canal = (
    df.groupby('canal_venda')
      .agg(Receita_Total=('receita', 'sum'),
           Lucro_Total=('lucro', 'sum'),
           Pedidos=('pedido_id', 'nunique'))
      .assign(Margem=lambda x: x['Lucro_Total'] / x['Receita_Total'],
              Ticket_Medio=lambda x: x['Receita_Total'] / x['Pedidos'])
      .reset_index()
)

def formata_moeda(v):
    return f'R$ {v:,.2f}'.replace(',', 'X').replace('.', ',').replace('X', '.')

resumo_executivo_view = resumo_executivo.copy()
resumo_executivo_view['Valor'] = resumo_executivo_view.apply(
    lambda row: f"{row['Valor']:.2%}".replace('.', ',') if row['KPI'] == 'Margem' else formata_moeda(row['Valor']),
    axis=1
)

resumo_canal_view = resumo_canal.copy()
resumo_canal_view['Receita_Total'] = resumo_canal_view['Receita_Total'].map(formata_moeda)
resumo_canal_view['Lucro_Total'] = resumo_canal_view['Lucro_Total'].map(formata_moeda)
resumo_canal_view['Margem'] = resumo_canal_view['Margem'].map(lambda v: f'{v:.2%}'.replace('.', ','))
resumo_canal_view['Ticket_Medio'] = resumo_canal_view['Ticket_Medio'].map(formata_moeda)

resumo_executivo_view, resumo_canal_view


(             KPI            Valor
 0  Receita Total  R$ 1.921.492,22
 1    Lucro Total    R$ 544.496,15
 2         Margem           28,34%
 3   Ticket Médio      R$ 4.574,98,
    canal_venda  Receita_Total    Lucro_Total  Pedidos  Margem Ticket_Medio
 0  Loja Física  R$ 640.797,06  R$ 174.583,16      143  27,24%  R$ 4.481,10
 1  Marketplace  R$ 657.802,09  R$ 185.802,57      141  28,25%  R$ 4.665,26
 2       Online  R$ 622.893,07  R$ 184.110,42      136  29,56%  R$ 4.580,10)

## 17. Entrega esperada

Seu notebook deve demonstrar:
- organização
- fórmulas corretas de KPI
- uso adequado de `groupby`, `agg`, `sort_values`
- leitura temporal coerente
- interpretação escrita dos resultados

### Revisão final da entrega
- KPIs globais calculados e interpretados.
- Tabelas por canal, categoria, UF e tempo concluídas.
- Rankings Top N de campeões e pontos de baixa eficiência concluídos.
- Conclusões executivas e decisão tática registradas.

### Mensagem principal da aula
Transformar linhas de dados limpos em **apoio real para decisões de negócio**.
